// ng-validate-equal/esm5/lib/ng-validate-equal.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { ValidateEqualDirective } from './ng-validate-equal.directive';
import * as ɵngcc0 from '@angular/core';
var ValidateEqualModule = /** @class */ (function () {
    function ValidateEqualModule() {
    }
ValidateEqualModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ValidateEqualModule });
ValidateEqualModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ValidateEqualModule_Factory(t) { return new (t || ValidateEqualModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ValidateEqualModule, { declarations: function () { return [ValidateEqualDirective]; }, exports: function () { return [ValidateEqualDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ValidateEqualModule, [{
        type: NgModule,
        args: [{
                declarations: [ValidateEqualDirective],
                imports: [],
                exports: [ValidateEqualDirective]
            }]
    }], function () { return []; }, null); })();
    return ValidateEqualModule;
}());
export { ValidateEqualModule };
