// ng2-ckeditor/esm2015/ckbutton.directive.js
import { Directive, EventEmitter, Output, Input } from '@angular/core';
/**
 * CKGroup component
 * Usage :
 *  <ckeditor [(ngModel)]="data" [config]="{...}" debounce="500">
 *      <ckbutton [name]="'SaveButton'" [command]="'saveCommand'" (click)="save($event)"
 *                [icon]="'/save.png'" [toolbar]="'customGroup,1'" [label]="'Save'">
 *      </ckbutton>
 *   </ckeditor>
 */
import * as ɵngcc0 from '@angular/core';
export class CKButtonDirective {
    constructor() {
        this.click = new EventEmitter();
    }
    initialize(editor) {
        editor.instance.addCommand(this.command, {
            exec: (evt) => {
                this.click.emit(evt);
            },
        });
        editor.instance.ui.addButton(this.name, {
            label: this.label,
            command: this.command,
            toolbar: this.toolbar,
            icon: this.icon,
        });
    }
    ngOnInit() {
        if (!this.name)
            throw new Error('Attribute "name" is required on <ckbutton>');
        if (!this.command)
            throw new Error('Attribute "command" is required on <ckbutton>');
    }
}
CKButtonDirective.ɵfac = function CKButtonDirective_Factory(t) { return new (t || CKButtonDirective)(); };
CKButtonDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: CKButtonDirective, selectors: [["ckbutton"]], inputs: { label: "label", command: "command", toolbar: "toolbar", name: "name", icon: "icon" }, outputs: { click: "click" } });
CKButtonDirective.propDecorators = {
    click: [{ type: Output }],
    label: [{ type: Input }],
    command: [{ type: Input }],
    toolbar: [{ type: Input }],
    name: [{ type: Input }],
    icon: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CKButtonDirective, [{
        type: Directive,
        args: [{
                selector: 'ckbutton'
            }]
    }], function () { return []; }, { click: [{
            type: Output
        }], label: [{
            type: Input
        }], command: [{
            type: Input
        }], toolbar: [{
            type: Input
        }], name: [{
            type: Input
        }], icon: [{
            type: Input
        }] }); })();
